### ISSUE TITLE
Tokens always pass validation even with invalid checksums when using asymmetric key algorithms

### ISSUE THREAD

**Issue by dlouzan** [SkyLothar/lua-resty-jwt#53]
Hello,

First of all, let me say that my knowledge of Lua can be measured in "days", so please be gentle :-)

I have been using the library for the last weeks, and I it seems that there is a critical bug in the validation of tokens when using asymmetric algorithms (RS256 or RS512).

In the following code in jwt.lua `_M:verify_jwt_object`:

``` lua
...
    if alg == str_const.RS256 then
      local verified, err = verifier:verify(message, sig, evp.CONST.SHA256_DIGEST)
    elseif alg == str_const.RS512 then
      local verified, err = verifier:verify(message, sig, evp.CONST.SHA512_DIGEST)
    end
    if not verified then
      jwt_obj[str_const.reason] = err
    end
  else
    jwt_obj[str_const.reason] = "Unsupported algorithm " .. alg
  end

  if not jwt_obj[str_const.reason] then
    jwt_obj[str_const.verified] = true
    jwt_obj[str_const.reason] = str_const.everything_awesome
  end
  return jwt_obj
...
```

As far as I understand, this piece of code creates variables `verified` and `err` in local block scope, meaning that out of the if-then-else they will be `nil`, no matter what happens after the verification. Then the next piece of code will set the result `verified` to true depending on `reason`, which was assigned to nil from the `err` value.

The end result seems to be that, no matter which key is passed, when this code path is reached the verification will always succeed.

I did a quick fix on my code with something like the following and it seems to work (please excuse my poor Lua):

``` lua
    -- a previous check makes sure that 'alg' can only be either 'str_const.RS256'
    -- or 'str_const.RS512'
    local evp_digest = nil
    if alg == str_const.RS256 then
      evp_digest = evp.CONST.SHA256_DIGEST
    else
      evp_digest = evp.CONST.SHA512_DIGEST
    end

    local verified, err = verifier:verify(message, sig, evp_digest)

    if not verified then
      jwt_obj[str_const.reason] = err
    end
  else 
    jwt_obj[str_const.reason] = "Unsupported algorithm " .. alg
  end

  if not jwt_obj[str_const.reason] then
    jwt_obj[str_const.verified] = true
    jwt_obj[str_const.reason] = str_const.everything_awesome
  end
  return jwt_obj
```

Am I missing anything here?

**Comment by zandbelt** [comment]
Seems like that was introduced only recently when adding support for RS512: https://github.com/SkyLothar/lua-resty-jwt/commit/6de92897ec6fab063d889b33baf6600df3b6f942

**Comment by SkyLothar** [comment]
I'll dig into it in few days. Thank you guys for reporting.

**Comment by lucafavatella** [comment]
I understand this shall be closed following #54.

**Comment by dlouzan** [comment]
Hello team, could we expect a release soon? I'm still fixing this manually in our build, and in the meantime another critical issue #57 was reported. Both seem to have been fixed in master, just expecting a release so that I can update our build configuration :-)

**Comment by SkyLothar** [comment]
Just released v0.1.10 on
- [github](https://github.com/SkyLothar/lua-resty-jwt/releases/tag/v0.1.10)
- [luarocks](https://luarocks.org/modules/skylothar/lua-resty-jwt)
- [opm](https://opm.openresty.org/)
